massa-versioning-worker/src/versioning.rs

```diff
@@ -0,0 +1,1057 @@
+    BeforeInitialState(MipStateTypeId, MassaTime),
+    #[error("Empty history, should never happen")]
+    EmptyHistory,
+    #[error("Cannot predict in the future (~ threshold not reach yet)")]
+    Unpredictable,
+}
+
+// Store
+
+/// Database for all MIP info
+#[derive(Debug, Clone)]
+pub struct MipStore(pub Arc<RwLock<MipStoreRaw>>);
+
+/*
+impl Default for MipStore {
+    fn default() -> Self {
+        Self(Arc::new(RwLock::new(Default::default())))
+    }
+}
+*/
+
+impl MipStore {
+    // TODO: merge get_version_current / get_version_to_announce so we don't iter twice?
+    //       & rename to get_network_versions(...) -> (u32, u32)
+    //       Move all others funcs from Factory and named them like:
+    //       get_component_version_XXX (best, all, ...)
+
+    /// Retrieve the current "global" version to set in block header
+    pub fn get_version_current(&self) -> u32 {
+        let lock = self.0.read();
+        let store = lock.deref();
+        // Current version == last active
+        store
+            .0
+            .iter()
+            .rev()
+            .find_map(|(k, v)| (v.state == MipState::active()).then_some(k.version))
+            .unwrap_or(0)
+    }
+
+    /// Retrieve the "global" version number to announce in block header
+    /// return 0 is there is nothing to announce
+    pub fn get_version_to_announce(&self) -> u32 {
+        let lock = self.0.read();
+        let store = lock.deref();
+        // Announce the latest versioning info in Started / LockedIn state
+        // Defined == Not yet ready to announce
+        // Active == current version
+        store
+            .0
+            .iter()
+            .rev()
+            .find_map(|(k, v)| {
+                matches!(&v.state, &MipState::Started(_) | &MipState::LockedIn(_))
+                    .then_some(k.version)
+            })
+            .unwrap_or(0)
+    }
+}
+
+impl<const N: usize> TryFrom<[(MipInfo, MipStateHistory); N]> for MipStore {
+    type Error = ();
+
+    fn try_from(value: [(MipInfo, MipStateHistory); N]) -> Result<Self, Self::Error> {
+        MipStoreRaw::try_from(value)
+            .and_then(|store_raw| Ok(Self(Arc::new(RwLock::new(store_raw)))))
+    }
+}
+
+/// Store of all versioning info
+#[derive(Debug, Clone, Default)]
+pub struct MipStoreRaw(pub(crate) BTreeMap<MipInfo, MipStateHistory>);
+
+impl MipStoreRaw {
+    /// Update our store with another (usually after a bootstrap where we received another store)
+    /// Return true if update is successful, false if something is wrong on given store raw
+    fn update_with(&mut self, store_raw: &MipStoreRaw) -> bool {
+        // iter over items in given store:
+        // -> 2 cases: VersioningInfo is already in self store -> add to 'to_update' list
+        //             VersioningInfo is not in self.store -> add to 'to_add' list
+        //
+
+        // TODO: Check component_version always increase...
```

Comment:
Couldn't we want to have a new MIP that reverts a component version? Why is that a problem?

Reply:
From what I see, this is prone to error (for user & dev) so I'd vote to disallow it. In Versioning followup, there is a proposal to add another state like Disabled in order to disable a version.
Does it sound ok to you?

Reply:
Yeah ok, we may need to think a bit more about that workflow in the follow up I think, but your reasoning does make sense.

Reply:
fixed